### PR TITLE
Sort transformation

### PR DESCRIPTION
**Pull request by nerik**
Adds support for `metadata.group` in GL layers generated by the LayerComposer. 
Layers will be sorted according to <a href="https://github.com/GlobalFishingWatch/map-styler/compare/sort?expand=1#diff-598de0b0c83d72ffa2d2a359634401ddR4">this order</a>, no matter what is the order of generators configs, adding the possibility for a generator to produce layers at distinct positions (ie basemap background + labels).
Additionally, support is provided for `mapbox:group` param used in legacy `map-client`s `style.json`, which will allow us to use `sort` for both legacy map-client layers and LayerComposer layers at the same time.

`sort` is a style transformation. Support for multiple transformations has been added in map-components: https://github.com/GlobalFishingWatch/map-components/pull/103

Transformations to be implemented:
- sort (GL style -> GL style with sorted layers)
- addInteraction (GL style -> GL style with selected features, selected features info, interactiveLayerIds)
- addAttributions (GL style ->GL style with concatenated attributions)
- addLegends (GL style -> with concatenated legends)

### CODE REVIEW

File: src/layer-composer/generators/background/background.ts [hunk 1]
```diff
@@ -1,16 +1,13 @@
+import { BackgroundPaint, BackgroundLayout } from 'mapbox-gl'
+import { ExtendedLayer, Group } from '../../../types/types'
```
Comment (j8seangel) [inline src/layer-composer/generators/background/background.ts:2]: There is support for absolute imports so I think worth it to use it and avoid hard to read imports
```suggestion
import { ExtendedLayer, Group } from 'types/types'
```

Also maybe a more specific name for the file helps as `types` sounds too generic

File: src/layer-composer/generators/heatmap/heatmap.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ import flatten from 'lodash/flatten'
 import compact from 'lodash/compact'
 import debounce from 'lodash/debounce'
 import zip from 'lodash/zip'
+import { Group } from '../../../types/types'
```
Comment (j8seangel) [inline src/layer-composer/generators/heatmap/heatmap.ts:5]: Same than above 🙏

File: src/sort/sort.test.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      },
+    },
+    {
+      id: 'custom',
+    },
+  ],
+}
+
+test('sort according to order', async () => {
+  const sortedStyle = sort(convertLegacyGroups(mock))
```
Comment (j8seangel) [inline src/sort/sort.test.ts:51]: 💘

File: src/sort/sort.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    })
+  return newStyle
+}
+
+const sort = (style: ExtendedStyle): ExtendedStyle => {
+  const newStyle = { ...style }
+  newStyle.layers = newStyle.layers?.sort((a: ExtendedLayer, b: ExtendedLayer) => {
+    const aGroup = a.metadata?.group || Group.Default
+    const bGroup = b.metadata?.group || Group.Default
+    const aPos = GROUP_ORDER.indexOf(aGroup)
```
Comment (j8seangel) [inline src/sort/sort.ts:45]: This means the GROUP_ORDER is fixed, isn't it? what about use it as default selection but also export the list to be able to customise from outside ?

File: src/sort/sort.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      const legacyGroup = layer.metadata && layer.metadata['mapbox:group']
+      if (legacyGroup) {
+        layer.metadata.group = GROUP_FROM_LEGACY[legacyGroup]
+      }
+      return layer
+    })
+  return newStyle
+}
+
+const sort = (style: ExtendedStyle): ExtendedStyle => {
```
Comment (j8seangel) [inline src/sort/sort.ts:40]: Something like:
```suggestion
const sort = (style: ExtendedStyle, groupOrder = GROUP_ORDER): ExtendedStyle => {
```

File: src/types/types.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
+import { Style, Layer } from 'mapbox-gl'
+
+export interface Dictionary<T> {
```
Comment (j8seangel) [inline src/types/types.ts:3]: Thanks, I've repeated this so many times!

**Review by j8seangel — CHANGES_REQUESTED** [review]
big 👏 ! I love the idea of make the transformations a chain and being able to do as many as needed extending the funcionality, great job!

Some more ideas to do in the future:
- [ ] Allow code splitting transformations importing from 'map-styler/transformations/sort` 
- [ ] Create transformationHook to compose with useMapStyler or to use independently 
- [ ] Customizable order

File: src/layer-composer/generators/background/background.ts (hunk 1 above)
Comment (nerik) [reply src/layer-composer/generators/background/background.ts:2]: I agree but that doesn't work :(
I have changed `types/types` to just `types` though

File: src/sort/sort.ts (hunk 4 above)
Comment (nerik) [reply src/sort/sort.ts:45]: It's a good idea but I'd prefer waiting for the use case to present itself before implementing.

**Pull request merged**